### PR TITLE
[feat]: Add configuration for per-route ratelimitting.

### PR DESCRIPTION
**Pull request by mikeykhalil**
Building off the work from this PR https://github.com/dollarshaveclub/guardian/pull/31. 
___
**Proposal**

In order to have a mechanism for ratelimiting on a per-route basis, we need a way to configure the ratelimits.

While it would technically be possible to configure route rate limits via command line arguments, this seems unrealistic in practice due to the complex data structure required to describe a route rate limit. 

I propose that we allow people to configure the route rate limits by specifying a path to a config file.

```
guardian-cli --redis-address localhost:6379 set-route-rate-limits ./config.yaml
```

The config file would be structured like so: 
```
route_rate_limits:
- route: "/foo/bar"
  limit:
    duration: 1m
    enabled: true
    count: 10
- route: "/foo/baz"
  limit:
    duration: 1m
    enabled: false
    count: 5
```

Then to remove a rate limit, you would do 

```
guardian-cli --redis-address localhost:6379 remove-route-rate-limits "/foo/bar,/foo/baz"
```

In its current state, the implementation does not provide a way to specify the configuration at the start of the server. Instead, it is only done through the guardian-cli and technically the server will not be configured until it syncs up with the redis conf store. 

The only reason I did this is because I do not quite understand the benefits of providing an initial configuration if it's going to get overwritten shortly after starting.

### CODE REVIEW

**Comment by bkeroackdsc** [comment]
I'm not the original author but if I recall correctly, the argument behind using Redis as a configuration store was twofold:

1. Redis is already in use as a data store backend for rate limiting
1. Redis would potentially allow other users to manipulate the configuration more easily (for example, application teams that don't have write access to the namespace that Guardian runs within)

The core of this proposal seems fine to me, but I would explore storing the proposed datastore in Redis alongside the existing configuration, primarily for consistency. We can argue whether we want to continue using Redis as a config store, but that change is of a wider scope than should be considered for per-route rate limiting.

**Comment by mikeykhalil** [comment]
[Guardian Per-Route Ratelimit](https://trello.com/c/Tidkq7OG/815-guardian-per-route-ratelimit)

File: cmd/guardian-cli/main.go [hunk 1]
```diff
@@ -122,6 +133,38 @@ func main() {
 			os.Exit(1)
 		}
 		fmt.Printf("%v\n", limit)
+	case getRouteRateLimitsCmd.FullCommand():
+		routeRateLimits, err := getRouteRateLimits(redisConfStore)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error getting route rate limits: %v\n", err)
```
Comment (bkeroackdsc) [inline cmd/guardian-cli/main.go]: This repeated error handling can go into a helper function:
```go
func fatalerror(err error) {
   fmt.Fprintf(os.Stderr, err.Error())
   os.Exit(1)
}
```

File: cmd/guardian-cli/main.go [hunk 2]
```diff
@@ -122,6 +133,38 @@ func main() {
...
+		for url, limit := range routeRateLimits {
+			entry := guardian.RouteRateLimitConfigEntry{
+				Route: url.EscapedPath(),
+				Limit: limit,
+			}
+			config.RouteRatelimits = append(config.RouteRatelimits, entry)
+		}
+		configYaml, err := yaml.Marshal(config)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error getting route rate limits: %v\n", err)
```
Comment (bkeroackdsc) [inline cmd/guardian-cli/main.go]: "error marshaling route limit yaml"

**Comment by mikeykhalil** [comment]
Addressed the comments and did some more testing. Think this is about ready. Will do a bit more testing next week.

**Comment by georgasa** [comment]
Hello,
I tried the above enhancement but didn't manage to make it working :(

./guardian-cli --redis-address localhost:6379 set-route-rate-limits ./guardian_config.yaml
guardian-cli: error: expected command but got "set-route-rate-limits", try --help

(app-root) sh-4.2$ cat guardian_config.yaml 
route_rate_limits:
- route: "/foo/bar"
  limit:
    duration: 1m
    enabled: true
    count: 10
- route: "/foo/baz"
  limit:
    duration: 1m
    enabled: false
    count: 5

thanks

**Comment by mikeykhalil** [comment]
> Hello,
> I tried the above enhancement but didn't manage to make it working :(
> 
> ./guardian-cli --redis-address localhost:6379 set-route-rate-limits ./guardian_config.yaml
> guardian-cli: error: expected command but got "set-route-rate-limits", try --help
> 
> (app-root) sh-4.2$ cat guardian_config.yaml
> route_rate_limits:
> 
> * route: "/foo/bar"
>   limit:
>   duration: 1m
>   enabled: true
>   count: 10
> * route: "/foo/baz"
>   limit:
>   duration: 1m
>   enabled: false
>   count: 5
> 
> thanks

Hey. Can you double check that you rebuilt the `guardian-cli` binary after checking out this branch? Note that you need to run `make cli` to build the cli tool. I switch between different versions of the cli and this error usually just pops up when using a version that doesn't actually have this command built in.

**Pull request merged**